### PR TITLE
Bugfix/issue_1343 foreground permissions exception

### LINKED ISSUE #1343: Missing Foreground Permission causes fatal exception
### Bug Report
If the router service attempts to start in the foreground without the manifest permission it will cause a fatal exception that will crash the app.

##### Reproduction Steps
1. Create a Test app without foreground permissions
2. Run application and connect to manticore

##### Expected Behavior
Observe Fatal exception

##### Observed Behavior
Application should not crash and should handle exception gracefully

### PR DESCRIPTION
**Pull request by RHenigan**
Fixes #1343 

This PR is **[ready]** for review.

### Risk
This PR makes **[no]** API changes.

### Testing Plan
- [x] I have verified that I have not introduced new warnings in this PR (or explain why below)
- [x] I have run the unit tests with this PR
- [x] I have tested this PR against Core and verified behavior (if applicable, if not applicable, explain why below).
- [x] I have tested Android, Java SE, and Java EE

### Test Steps
 - Comment out FOREGROUND_SERVICE permissions
 - Try to run the app connecting to TDK through Bluetooth
 - Ensure app is able to run without crashing

### Summary
This PR will ensure that permissions are checked before calling startForeground and catch exceptions thrown by startForeground

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: android/hello_sdl_android/src/main/java/com/sdl/hellosdlandroid/SdlReceiver.java [hunk 1]
```diff
@@ -1,12 +1,15 @@
...
+import android.Manifest;
 import android.content.Context;
 import android.content.Intent;
 import android.os.Build;
 import android.util.Log;
 
 import com.smartdevicelink.transport.SdlBroadcastReceiver;
 import com.smartdevicelink.transport.SdlRouterService;
+import com.smartdevicelink.util.AndroidTools;
+import com.smartdevicelink.util.DebugTool;
```
[the remark this answers is not in the document]
Comment (RHenigan) [reply android/hello_sdl_android/src/main/java/com/sdl/hellosdlandroid/SdlReceiver.java]: updated

File: android/sdl_android/src/main/java/com/smartdevicelink/transport/SdlRouterService.java [hunk 2]
```diff
@@ -1484,37 +1484,14 @@ private void enterForeground(String content, long chronometerLength, boolean ong
...
-			safeStartForeground(FOREGROUND_SERVICE_ID, notification);
+
+			safeStartForeground(FOREGROUND_SERVICE_ID, builder.build());
 			isForeground = true;
 
 		}
  
     }
 
-	/**
```
[the remark this answers is not in the document]
Comment (RHenigan) [reply android/sdl_android/src/main/java/com/smartdevicelink/transport/SdlRouterService.java]: Moved back, I had put it back to the wrong place after I had originally gotten rid of it.

File: android/sdl_android/src/main/java/com/smartdevicelink/transport/SdlRouterService.java [hunk 3]
```diff
@@ -1484,37 +1484,14 @@ private void enterForeground(String content, long chronometerLength, boolean ong
 				}
 				return;
 			}
-			safeStartForeground(FOREGROUND_SERVICE_ID, notification);
+
+			safeStartForeground(FOREGROUND_SERVICE_ID, builder.build());
```
[the remark this answers is not in the document]
Comment (RHenigan) [reply android/sdl_android/src/main/java/com/smartdevicelink/transport/SdlRouterService.java]: This was a mistake, thanks for catching it

File: android/hello_sdl_android/src/main/java/com/sdl/hellosdlandroid/SdlService.java [hunk 4]
```diff
@@ -6,6 +6,7 @@
 import android.app.NotificationManager;
 import android.app.Service;
 import android.content.Context;
+import android.content.ContextWrapper;
```
[the remark this answers is not in the document]
Comment (RHenigan) [reply android/hello_sdl_android/src/main/java/com/sdl/hellosdlandroid/SdlService.java]: updated

File: android/sdl_android/src/main/java/com/smartdevicelink/util/AndroidTools.java [hunk 5]
```diff
@@ -44,6 +47,8 @@
 import android.graphics.Bitmap;
 import android.graphics.BitmapFactory;
 import android.os.BatteryManager;
+import android.os.Build;
+import android.support.v4.app.NotificationCompat;
```
[the remark this answers is not in the document]
Comment (RHenigan) [reply android/sdl_android/src/main/java/com/smartdevicelink/util/AndroidTools.java]: updated

File: android/sdl_android/src/main/java/com/smartdevicelink/util/AndroidTools.java [hunk 6]
```diff
@@ -57,7 +62,12 @@
 import java.util.HashMap;
 import java.util.List;
 
+import static com.smartdevicelink.transport.TransportConstants.SDL_NOTIFICATION_CHANNEL_ID;
+
 public class AndroidTools {
+
+	public static final String SDL_NOTIFICATION_CHANNEL_ID 						= "sdl_notification_channel";
```
[the remark this answers is not in the document]
Comment (RHenigan) [reply android/sdl_android/src/main/java/com/smartdevicelink/util/AndroidTools.java]: updated

File: android/sdl_android/src/main/java/com/smartdevicelink/util/AndroidTools.java [hunk 7]
```diff
@@ -57,7 +62,12 @@
 import java.util.HashMap;
 import java.util.List;
 
+import static com.smartdevicelink.transport.TransportConstants.SDL_NOTIFICATION_CHANNEL_ID;
```
[the remark this answers is not in the document]
Comment (RHenigan) [reply android/sdl_android/src/main/java/com/smartdevicelink/util/AndroidTools.java]: updated

File: android/sdl_android/src/main/java/com/smartdevicelink/transport/SdlRouterStatusProvider.java [hunk 8]
```diff
@@ -134,8 +136,12 @@ private boolean bindToService(){
...
 			bindingIntent.putExtra(FOREGROUND_EXTRA, true);
 			SdlBroadcastReceiver.setForegroundExceptionHandler(); //Prevent ANR in case the OS takes too long to start the service
-			context.startForegroundService(bindingIntent);
-
+			int permission = context.checkPermission(Manifest.permission.FOREGROUND_SERVICE, android.os.Process.myPid(), android.os.Process.myUid());
+			if (permission != -1) {
+				AndroidTools.safeStartForegroundService(context, bindingIntent);
+			} else {
+				DebugTool.logError("Foreground Permissions Not Enabled: will not start foreground permissions");
+			}
```
[the remark this answers is not in the document]
Comment (RHenigan) [reply android/sdl_android/src/main/java/com/smartdevicelink/transport/SdlRouterStatusProvider.java]: updated

File: android/sdl_android/src/main/java/com/smartdevicelink/util/AndroidTools.java [hunk 9]
```diff
@@ -191,4 +201,60 @@ public static Bitmap downloadImage(String urlStr) throws IOException {
...
+	 * is null, or a notification was unable to be created we will still attempt to call the
+	 * startForeground method in hopes that Android will not throw the System Exception. In the
+	 * event that the user is on Android 28+ it will check the FOREGROUND_SERVICE permissions
+	 * before trying to call startForeground.
+	 * @param service a service instance
+	 * @param context a context instance
+	 * @param id notification channel id
+	 * @param notification the notification to display when in the foreground
+	 */
+	public static void safeStartForeground(Service service, Context context, int id, Notification notification){
```
[the remark this answers is not in the document]
Comment (RHenigan) [reply android/sdl_android/src/main/java/com/smartdevicelink/util/AndroidTools.java]: removed

[the remark this answers is not in the document]
Comment (RHenigan) [reply android/sdl_android/src/main/java/com/smartdevicelink/util/AndroidTools.java]: to investigate if this logic is needed in SdlRouterService implementation

File: android/hello_sdl_android/src/main/java/com/sdl/hellosdlandroid/SdlService.java [hunk 10]
```diff
@@ -105,7 +107,11 @@ public void enterForeground() {
 						.setContentTitle("Connected through SDL")
 						.setSmallIcon(R.drawable.ic_sdl)
 						.build();
-				startForeground(FOREGROUND_SERVICE_ID, serviceNotification);
+				try {
```
[the remark this answers is not in the document]
Comment (RHenigan) [reply android/hello_sdl_android/src/main/java/com/sdl/hellosdlandroid/SdlService.java]: updated

File: android/sdl_android/src/main/java/com/smartdevicelink/util/AndroidTools.java [hunk 11]
```diff
@@ -191,4 +201,60 @@ public static Bitmap downloadImage(String urlStr) throws IOException {
...
+	/**
+	 * This is a wrapper around the startForegroundService method. In the
+	 * event that the user is on Android 28+ it will check the FOREGROUND_SERVICE permissions
+	 * before trying to call startForegroundService.
+	 * @param context a context instance
+	 * @param intent the foreground service intent
+	 * @return a ComponentName, an identifier for the started service, will return null is service
+	 * was unable to start
+	 */
+	public static ComponentName safeStartForegroundService(Context context, Intent intent){
```
[the remark this answers is not in the document]
Comment (RHenigan) [reply android/sdl_android/src/main/java/com/smartdevicelink/util/AndroidTools.java]: Changes made

File: android/sdl_android/src/main/java/com/smartdevicelink/util/AndroidTools.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (RHenigan) [reply android/sdl_android/src/main/java/com/smartdevicelink/util/AndroidTools.java]: According to android documentation ([startForeground](https://developer.android.com/reference/android/app/Service#startForeground(int,%20android.app.Notification))

File: android/sdl_android/src/main/java/com/smartdevicelink/util/AndroidTools.java [hunk 12]
```diff
@@ -191,4 +196,31 @@ public static Bitmap downloadImage(String urlStr) throws IOException {
...
+			if (inDebugMode || permissionGranted) {
+				return context.startForegroundService(intent);
+			} else {
+				DebugTool.logError("App missing FOREGROUND_SERVICE Permissions");
+			}
+		} else if (Build.VERSION.SDK_INT == Build.VERSION_CODES.O) {
+			return context.startForegroundService(intent);
+		} else if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
+			DebugTool.logError("startForegroundService() call requires Android Oreo or newer");
+		}
```
Comment (bilal-alsharifi) [inline android/sdl_android/src/main/java/com/smartdevicelink/util/AndroidTools.java:223]: I have just learned that there are two versions of Android O (`VERSION_CODES.O`= `26` & `VERSION_CODES.O_MR1` = `27`) and they both don't require the foreground permission. So I updated the code accordingly 

```suggestion
		} else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
			return context.startForegroundService(intent);
		}
```

File: android/sdl_android/src/main/java/com/smartdevicelink/util/AndroidTools.java [hunk 13]
```diff
@@ -191,4 +196,29 @@ public static Bitmap downloadImage(String urlStr) throws IOException {
...
+	 */
+	@RequiresApi(api = Build.VERSION_CODES.O)
+	public static ComponentName safeStartForegroundService(Context context, Intent intent) {
+		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
+			boolean inDebugMode = (0 != (context.getApplicationInfo().flags & ApplicationInfo.FLAG_DEBUGGABLE));
+			boolean permissionGranted = PackageManager.PERMISSION_GRANTED == context.checkPermission(Manifest.permission.FOREGROUND_SERVICE, android.os.Process.myPid(), android.os.Process.myUid());
+			if (inDebugMode || permissionGranted) {
+				return context.startForegroundService(intent);
+			} else {
+				DebugTool.logError("App missing FOREGROUND_SERVICE Permissions");
```
Comment (bilal-alsharifi) [inline android/sdl_android/src/main/java/com/smartdevicelink/util/AndroidTools.java:217]:
```suggestion
				DebugTool.logError("App is missing FOREGROUND_SERVICE Permission. Sdl will not work.");
```

**Pull request merged**